File: src/main/java/pd/discordipc/models/RichPresence.java
package pd.discordipc.models;

import pd.discordipc.utils.TimeUtils;
import java.util.List;

public class RichPresence {
    private String details;
    private String state;
    private ActivityType type = ActivityType.PLAYING;
    private String url;
    private Long startTimestamp;
    private Long endTimestamp;
    private String largeImageKey;
    private String largeImageText;
    private String smallImageKey;
    private String smallImageText;
    private String partyId;
    private Integer partySize;
    private Integer partyMax;
    private String matchSecret;
    private String joinSecret;
    private String spectateSecret;
    private List<Button> buttons;

    public String getDetails() { return details; }
    public String getState() { return state; }
    public ActivityType getType() { return type; }
    public String getUrl() { return url; }
    public Long getStartTimestamp() { return startTimestamp; }
    public Long getEndTimestamp() { return endTimestamp; }
    public String getLargeImageKey() { return largeImageKey; }
    public String getLargeImageText() { return largeImageText; }
    public String getSmallImageKey() { return smallImageKey; }
    public String getSmallImageText() { return smallImageText; }
    public String getPartyId() { return partyId; }
    public Integer getPartySize() { return partySize; }
    public Integer getPartyMax() { return partyMax; }
    public String getMatchSecret() { return matchSecret; }
    public String getJoinSecret() { return joinSecret; }
    public String getSpectateSecret() { return spectateSecret; }
    public List<Button> getButtons() { return buttons; }

    private RichPresence() {}

    public static class Builder {
        private final RichPresence rp = new RichPresence();

        public Builder details(String details) {
            rp.details = details;
            return this;
        }

        public Builder state(String state) {
            rp.state = state;
            return this;
        }

        public Builder type(ActivityType type) {
            rp.type = type;
            return this;
        }

        public Builder streaming(String url) {
            rp.type = ActivityType.STREAMING;
            rp.url = url;
            return this;
        }

        public Builder timer(String time) {
            long now = System.currentTimeMillis() / 1000L;
            rp.startTimestamp = now - TimeUtils.parseToSeconds(time);
            rp.endTimestamp = null;
            return this;
        }

        public Builder loop(String min, String max) {
            long now = System.currentTimeMillis() / 1000L;
            long minSec = TimeUtils.parseToSeconds(min);
            long maxSec = TimeUtils.parseToSeconds(max);
            long diff = maxSec - minSec;
            if (diff > 0) {
                long offset = (now % diff) + minSec;
                rp.startTimestamp = now - offset;
                rp.endTimestamp = null;
            }
            return this;
        }

        public Builder large(String key) {
            rp.largeImageKey = key;
            return this;
        }

        public Builder large(String key, String text) {
            rp.largeImageKey = key;
            rp.largeImageText = text;
            return this;
        }

        public Builder small(String key) {
            rp.smallImageKey = key;
            return this;
        }

        public Builder small(String key, String text) {
            rp.smallImageKey = key;
            rp.smallImageText = text;
            return this;
        }

        public Builder party(String id, int size, int max) {
            rp.partyId = id;
            rp.partySize = size;
            rp.partyMax = max;
            return this;
        }

        public Builder buttons(List<Button> buttons) {
            rp.buttons = buttons;
            return this;
        }

        public Builder button(String label, String url) {
            if (rp.buttons == null) rp.buttons = new java.util.ArrayList<>();
            rp.buttons.add(new Button(label, url));
            return this;
        }

        public RichPresence build() {
            return rp;
        }
    }

    public static class Button {
        private final String label;
        private final String url;

        public Button(String label, String url) {
            this.label = label;
            this.url = url;
        }

        public String getLabel() { return label; }
        public String getUrl() { return url; }
    }
}

File: src/main/kotlin/com/pdlayer/discordipc/models/RichPresence.kt
package com.pdlayer.discordipc.models

import com.pdlayer.discordipc.utils.TimeUtils
import java.net.URI

data class RichPresence(
    val details: String? = null,
    val state: String? = null,
    val type: ActivityType = ActivityType.PLAYING,
    val url: String? = null,
    val timestamps: Timestamps? = null,
    val assets: Assets? = null,
    val party: Party? = null,
    val secrets: Secrets? = null,
    val buttons: List<Button>? = null,
    val detailsUrl: String? = null,
    val stateUrl: String? = null,
    val instance: Boolean = false,
) {
    init {
        require(buttons == null || buttons.size <= 2) {
            "Discord Rich Presence supports at most 2 buttons"
        }
        requireHttpUrl(detailsUrl, "Details url")
        requireHttpUrl(stateUrl, "State url")
    }

    data class Timestamps(
        val start: Long? = null,
        val end: Long? = null,
    ) {
        companion object {
            fun elapsed(time: String): Timestamps {
                val now = System.currentTimeMillis() / 1000L
                return Timestamps(start = now - TimeUtils.parseToSeconds(time))
            }

            fun loop(min: String, max: String): Timestamps {
                val now = System.currentTimeMillis() / 1000L
                val minSeconds = TimeUtils.parseToSeconds(min)
                val maxSeconds = TimeUtils.parseToSeconds(max)
                val diff = maxSeconds - minSeconds

                require(diff > 0) { "Loop max ($max) must be greater than loop min ($min)" }

                val offset = (now % diff) + minSeconds
                return Timestamps(start = now - offset)
            }
        }
    }

    data class Assets(
        val largeImageKey: String? = null,
        val largeImageText: String? = null,
        val smallImageKey: String? = null,
        val smallImageText: String? = null,
        val largeImageUrl: String? = null,
        val smallImageUrl: String? = null,
    ) {
        init {
            requireHttpUrl(largeImageUrl, "Large image url")
            requireHttpUrl(smallImageUrl, "Small image url")
        }
    }

    data class Party(
        val id: String,
        val size: Int,
        val max: Int,
    ) {
        init {
            require(id.isNotBlank()) { "Party id must not be blank" }
            require(size >= 0) { "Party size cannot be negative: $size" }
            require(max >= 1) { "Party max must be at least 1: $max" }
            require(size <= max) {
                "Party size cannot be greater than party max"
            }
        }
    }

    data class Secrets(
        val match: String? = null,
        val join: String? = null,
        val spectate: String? = null,
    )

    data class Button(
        val label: String,
        val url: String,
    ) {
        init {
            require(label.isNotBlank()) { "Button label must not be blank" }
            require(label.length <= MAX_LABEL_LENGTH) {
                "Button label is longer than $MAX_LABEL_LENGTH characters: \"$label\""
            }
            require(isHttpUrl(url)) { "Button url must be a valid http(s) URL: \"$url\"" }
            require(url.length <= MAX_URL_LENGTH) {
                "Button url is longer than $MAX_URL_LENGTH characters"
            }
        }
    }

    companion object {
        fun playing(
            details: String? = null,
            state: String? = null,
            timestamps: Timestamps? = null,
            assets: Assets? = null,
            party: Party? = null,
            buttons: List<Button>? = null,
        ): RichPresence = activity(ActivityType.PLAYING, details, state, null, timestamps, assets, party, buttons)

        fun streaming(
            details: String? = null,
            state: String? = null,
            url: String,
            timestamps: Timestamps? = null,
            assets: Assets? = null,
            party: Party? = null,
            buttons: List<Button>? = null,
        ): RichPresence = activity(ActivityType.STREAMING, details, state, url, timestamps, assets, party, buttons)

        fun listening(
            details: String? = null,
            state: String? = null,
            timestamps: Timestamps? = null,
            assets: Assets? = null,
            party: Party? = null,
            buttons: List<Button>? = null,
        ): RichPresence = activity(ActivityType.LISTENING, details, state, null, timestamps, assets, party, buttons)

        fun watching(
            details: String? = null,
            state: String? = null,
            timestamps: Timestamps? = null,
            assets: Assets? = null,
            party: Party? = null,
            buttons: List<Button>? = null,
        ): RichPresence = activity(ActivityType.WATCHING, details, state, null, timestamps, assets, party, buttons)

        fun competing(
            details: String? = null,
            state: String? = null,
            timestamps: Timestamps? = null,
            assets: Assets? = null,
            party: Party? = null,
            buttons: List<Button>? = null,
        ): RichPresence = activity(ActivityType.COMPETING, details, state, null, timestamps, assets, party, buttons)

        fun custom(
            details: String? = null,
            state: String? = null,
            timestamps: Timestamps? = null,
            assets: Assets? = null,
            party: Party? = null,
            buttons: List<Button>? = null,
        ): RichPresence = activity(ActivityType.CUSTOM, details, state, null, timestamps, assets, party, buttons)

        private fun activity(
            type: ActivityType,
            details: String?,
            state: String?,
            url: String?,
            timestamps: Timestamps?,
            assets: Assets?,
            party: Party?,
            buttons: List<Button>?,
        ): RichPresence = RichPresence(
            details = details,
            state = state,
            type = type,
            url = url,
            timestamps = timestamps,
            assets = assets,
            party = party,
            buttons = buttons?.toList(),
        )
    }
}

private const val MAX_LABEL_LENGTH = 32

private const val MAX_URL_LENGTH = 512

private fun requireHttpUrl(value: String?, name: String) {
    if (value == null) return
    require(isHttpUrl(value)) { "$name must be a valid http(s) URL: \"$value\"" }
}

private fun isHttpUrl(value: String): Boolean = try {
    val uri = URI.create(value)
    val scheme = uri.scheme

    (scheme.equals("http", ignoreCase = true) || scheme.equals("https", ignoreCase = true)) &&
        !uri.host.isNullOrBlank()
} catch (_: Exception) {
    false
}
